apps/aecore/lib/aecore/util/state_trees.ex

```diff
@@ -0,0 +1,95 @@
+defmodule Aecore.Util.StateTrees do
+  alias Aecore.Account.AccountStateTree
+  alias Aecore.Contract.CallStateTree
+  alias Aecore.Contract.ContractStateTree
+  alias Aecore.Channel.ChannelStateTree
+  alias Aecore.Naming.NamingStateTree
+
+  @moduledoc """
+  Abstract module defining functions for known Tree types
+  """
+  defmacro __using__(_) do
+    quote location: :keep do
+      alias Aecore.Contract.{Contract, ContractStateTree}
+      alias Aecore.Chain.Identifier
+      alias Aecore.Naming.{Name, NameCommitment}
+      alias Aecore.Util.StateTrees
+      alias Aeutil.PatriciaMerkleTree
+      alias Aeutil.Serialization
+      alias MerklePatriciaTree.Trie
+
+      @typedoc "Hash of the tree"
+      @type hash :: binary()
+
+      @spec init_empty :: Trie.t()
+      def init_empty do
+        PatriciaMerkleTree.new(StateTrees.tree_type(__MODULE__))
+      end
+
+      @spec put(Trie.t(), binary(), map()) :: Trie.t()
+      def put(tree, key, value) do
```

Comment:
There is no type safety here. Imagine that we put an oracle into the channels trie and then try to fetch it.